fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,647 @@
+        return confirmed_block.body.execution_payload.block_hash
+    else:
+        return Hash32()
+```
+
+*Note*: This helper uses beacon block container extended in [Bellatrix](../specs/bellatrix/beacon-chain.md).
+
+## Confirmation Score
+
+The confirmation score for a block is the maximum adversarial percentage weight that a confirmed block can tolerate.
+This section specifies the algorithm to calculate the confirmation score of a given block.
+This under the assumption that the adversary is willing to get as much stake as possible slashed to
+prevent a block from being confirmed.
+
+### Helper Functions
+
+#### `get_one_confirmation_score`
+
+```python
+def get_one_confirmation_score(store: Store, block_root: Root) -> int:
+    current_slot = get_current_slot(store)
+    block = store.blocks[block_root]
+    parent_block = store.blocks[block.parent_root]
+    support = int(get_weight(store, block_root))
+    justified_state = store.checkpoint_states[store.justified_checkpoint]
+    maximum_support = int(
+        get_committee_weight_between_slots(justified_state, Slot(parent_block.slot + 1), current_slot)
+    )
+    proposer_score = int(get_proposer_score(store))
+
+    """
+    Return the max possible one_confirmation_score such that:
+    support / maximum_support > \
+        0.5 * (1 + proposer_score / maximum_support) + one_confirmation_score / 100
+    """
+    return max((100 * support - 50 * proposer_score - 1) // maximum_support - 50, -1)
+```
+
+#### `get_lmd_confirmation_score`
+
+```python
+def get_lmd_confirmation_score(store: Store, block_root: Root) -> int:
+    if block_root == store.finalized_checkpoint.root:
+        return 100 // 3
+    else:
+        block = store.blocks[block_root]
+        finalized_block = store.blocks[store.finalized_checkpoint.root]
+        if block.slot <= finalized_block.slot:
+            # This block is not in the finalized chain.
+            return -1
+        else:
+            # Check one_confirmed score for this block and LMD_confirmed score for the preceding chain.
+            return min(
+                get_one_confirmation_score(store, block_root),
+                get_lmd_confirmation_score(store, block.parent_root)
+            )
+```
+
+#### `get_ffg_confirmation_score_current_epoch`
+
+```python
+def get_ffg_confirmation_score_current_epoch(total_active_balance: int,
+                                             ffg_support_for_checkpoint: int,
+                                             min_ffg_support_slash_th: int,
+                                             remaining_ffg_weight: int) -> int:
+    assert min_ffg_support_slash_th <= ffg_support_for_checkpoint
+    
+    """
+    Return the max possible ffg_confirmation_score such that:
+    2 / 3 * total_active_balance <= \
+        ffg_support_for_checkpoint - \
+        min(min_ffg_support_slash_th, total_active_balance * ffg_confirmation_score / 100) + \
```

Comment:
Shouldn't it be?
```suggestion
        min(min_ffg_support_slash_th, (total_active_balance - remaining_ffg_weight) * ffg_confirmation_score / 100) + \
```

Reply:
Using `total_active_balances` rather than `total_active_balance - remaining_ffg_weight` is required in order to ensure monotonicity.
This deviates from the formula in the current public version of the paper. However, I expect that this modification will be incorporated in the next version of the paper.